src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -279,13 +303,20 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
     usable_free /= HeapWordSize;
     if ((free != usable_free) && (free - usable_free < ShenandoahHeap::min_fill_size() * HeapWordSize)) {
       // We'll have to add another card's memory to the padding
-      usable_free -= CardTable::card_size();
+      if (usable_free > CardTable::card_size_in_words()) {
+        usable_free -= CardTable::card_size_in_words();
+      } else {
+        assert(usable_free == 0, "usable_free is a multiple of card_size and card_size > min_fill_size");
+      }
     }
+    assert(size % CardTable::card_size_in_words() == 0, "PLAB size must be multiple of remembered set card size");
     if (size <= usable_free) {
-      assert(size % CardTable::card_size_in_words() == 0, "PLAB size must be multiple of remembered set card size");
-
       result = r->allocate_aligned(size, req, CardTable::card_size());
-      if (result != nullptr) {
+      assert(result != nullptr, "Allocation cannot fail");
+      assert(r->top() <= r->end(), "Allocation cannot span end of region");
+      assert(req.actual_size() % CardTable::card_size_in_words() == 0, "PLAB start must align with card boundary");
```

Comment:
This allows us to register objects in PLABs without acquiring a lock.  Otherwise, we need a lock because two threads might be registering objects within the same card in parallel.